### PR TITLE
tests: net: socket: tcp: Add test for big block data transfer

### PR DESCRIPTION
**Pull request by ssharks**
Add tests to reproduce the issues described in:
#44018
#44012

It send a data chunk from server to client over the loopback interface. It repeats the same, while introducing packet loss at the loopback driver.

### CODE REVIEW

File: drivers/net/Kconfig [hunk 1]
```diff
@@ -183,6 +183,12 @@ menuconfig NET_LOOPBACK
 
 if NET_LOOPBACK
 
+config NET_LOOPBACK_PACKET_DROP
+	bool "Controlable packet drop"
+	default n
```
Comment (rlubos) [inline drivers/net/Kconfig:188]: No need for `default n`, it's implicit.

File: drivers/net/loopback.c [hunk 2]
```diff
@@ -60,6 +60,15 @@ static void loopback_init(struct net_if *iface)
 	}
 }
 
+#ifdef CONFIG_NET_LOOPBACK_PACKET_DROP
+static float loopback_packet_drop_ratio = 0.0f;
+static float loopback_packet_drop_state = 0.0f;
+
+void loopback_set_packet_drop_ratio(float ratio) {
```
Comment (rlubos) [inline drivers/net/loopback.c:67]: The API function should be exposed in a header file. I see there's no `net/loopback.h` yet, I think it should be added.

File: drivers/net/loopback.c [hunk 3]
```diff
@@ -102,11 +111,27 @@ static int loopback_send(const struct device *dev, struct net_pkt *pkt)
...
+	loopback_packet_drop_state += loopback_packet_drop_ratio;
+	if (loopback_packet_drop_state < 1.0f) {
+#endif
+		res = net_recv_data(net_pkt_iface(cloned), cloned);
+		if (res < 0) {
+			LOG_ERR("Data receive failed.");
+		}
+#ifdef CONFIG_NET_LOOPBACK_PACKET_DROP
+	} else {
+		// administrate we dropped a packet
```
Comment (rlubos) [inline drivers/net/loopback.c:127]: Please use C-style comments.

File: drivers/net/loopback.c [hunk 4]
```diff
@@ -102,11 +111,27 @@ static int loopback_send(const struct device *dev, struct net_pkt *pkt)
...
-
-	res = net_recv_data(net_pkt_iface(cloned), cloned);
-	if (res < 0) {
-		LOG_ERR("Data receive failed.");
+	
+#ifdef CONFIG_NET_LOOPBACK_PACKET_DROP
+	/* Drop packets based on the loopback_packet_drop_ratio
+	a ratio of 0.2 will drop one every 5 packets */
+	loopback_packet_drop_state += loopback_packet_drop_ratio;
+	if (loopback_packet_drop_state < 1.0f) {
```
Comment (rlubos) [inline drivers/net/loopback.c:119]: If you reverse the logic here (`if (loopback_packet_drop_state  >= 1.0f) {`) you could have a single, cocise `#ifdef` block. It could end up with `goto out;` so you won't even need to modify existing code.

File: tests/net/socket/tcp/src/main.c [hunk 5]
```diff
@@ -234,15 +234,59 @@ void test_v6_send_recv(void)
 	k_sleep(TCP_TEARDOWN_TIMEOUT);
 }
 
-void test_v4_sendto_recvfrom(void)
+/* Test the stack behavior with a resonable sized block data, be sure to have multiple packets */
+#define TEST_BLOCK_SIZE 60000
+
+static uint8_t testBlockTcp[TEST_BLOCK_SIZE];
+static uint8_t testBlockTcpRx[TEST_BLOCK_SIZE + 10];
```
Comment (rlubos) [inline tests/net/socket/tcp/src/main.c:241]: This will greatly increase the RAM requirements of the test suite. Perhaps it'd be better to use smaller buffers (1k for example) and just re-send it several times?

Comment (rlubos) [reply tests/net/socket/tcp/src/main.c:241]: Also, please use snake_case for Zephyr code.

File: tests/net/socket/tcp/src/main.c [hunk 6]
```diff
@@ -234,15 +234,59 @@ void test_v6_send_recv(void)
...
+	while (total_received < TEST_BLOCK_SIZE) {
+		recved = recv(new_sock, testBlockTcpRx + total_received, sizeof(testBlockTcpRx) - total_received, 0);
+		
+		zassert(recved > 0, "received bigger then 0",
+			"Error receiving bytes %i bytes, got %i on top of %i in iteration %i, errno %i", TEST_BLOCK_SIZE,
+			recved, total_received, iteration, errno);
+		total_received += recved;
+		iteration++;
+	}
+	zassert_equal(memcmp(testBlockTcpRx, testBlockTcp, sizeof(testBlockTcp)),
```
Comment (rlubos) [inline tests/net/socket/tcp/src/main.c:272]: There's `zassert_mem_equal()` for this such checks.

File: tests/net/socket/tcp/src/main.c [hunk 7]
```diff
@@ -252,26 +296,90 @@ void test_v4_sendto_recvfrom(void)
...
-	test_sendto(c_sock, TEST_STR_SMALL, strlen(TEST_STR_SMALL), 0,
-		    (struct sockaddr *)&s_saddr, sizeof(s_saddr));
-
-	test_accept(s_sock, &new_sock, &addr, &addrlen);
-	zassert_equal(addrlen, sizeof(struct sockaddr_in), "wrong addrlen");
+	(void)k_thread_create(&tcp_server_thread_data, tcp_server_stack_area,
+	                                 K_THREAD_STACK_SIZEOF(tcp_server_stack_area),
+	                                 tcp_server_block_thread,
+	                                 &s_sock, NULL, NULL,
+	                                 k_thread_priority_get( k_current_get()), 0, K_NO_WAIT);
```
Comment (rlubos) [inline tests/net/socket/tcp/src/main.c:303]: Please fix the alignment (align with the opening bracket).

File: tests/net/socket/tcp/src/main.c [hunk 8]
```diff
@@ -252,26 +296,90 @@ void test_v4_sendto_recvfrom(void)
...
 
-	test_recvfrom(new_sock, MSG_PEEK, &addr, &addrlen);
-	zassert_equal(addrlen, sizeof(struct sockaddr_in), "wrong addrlen");
+	test_connect(c_sock, (struct sockaddr *)&s_saddr, sizeof(s_saddr));
+	LOG_ERR("Test this log message");
```
Comment (rlubos) [inline tests/net/socket/tcp/src/main.c:306]: I don't think it's needed in the test.

File: tests/net/socket/tcp/src/main.c [hunk 9]
```diff
@@ -234,15 +234,59 @@ void test_v6_send_recv(void)
...
+		iteration++;
+	}
+	zassert_equal(memcmp(testBlockTcpRx, testBlockTcp, sizeof(testBlockTcp)),
+		0,
+		"unexpected data");
+	
+	test_close(new_sock);
+}
+
+void test_v4_send_recv_block(void)
```
Comment (rlubos) [inline tests/net/socket/tcp/src/main.c:279]: I wonder about the name here, `block` suggests that the test is about the blocking socket, while it's not essential here. Maybe rename it to `test_v4_send_recv_large`?

File: tests/net/socket/tcp/src/main.c [hunk 10]
```diff
@@ -1075,7 +1183,10 @@ void test_main(void)
 		ztest_unit_test(test_v6_so_rcvtimeo),
 		ztest_unit_test(test_v4_msg_waitall),
 		ztest_unit_test(test_v6_msg_waitall),
-		ztest_user_unit_test(test_socket_permission)
+		ztest_user_unit_test(test_socket_permission),
+		ztest_unit_test(test_v4_send_recv_block),
+		ztest_unit_test_setup_teardown(test_v4_send_recv_block,
```
Comment (rlubos) [inline tests/net/socket/tcp/src/main.c:1188]: The test is added twice, is that intended?

Comment (ssharks) [reply tests/net/socket/tcp/src/main.c:1188]: Yes, the first one is without packet loss, the second one with

File: drivers/net/loopback.c (hunk 3 above)
Comment (ssharks) [reply drivers/net/loopback.c]: Done

File: drivers/net/loopback.c (hunk 4 above)
Comment (ssharks) [reply drivers/net/loopback.c]: Thanks for the suggestion, done

File: tests/net/socket/tcp/src/main.c [hunk 11]
```diff
@@ -234,15 +234,59 @@ void test_v6_send_recv(void)
...
 	int c_sock;
 	int s_sock;
-	int new_sock;
 	struct sockaddr_in c_saddr;
 	struct sockaddr_in s_saddr;
-	struct sockaddr addr;
-	socklen_t addrlen = sizeof(addr);
+
+	/* initialize the test block with easily debuggable data */
+	for (uint32_t i=0; i<TEST_BLOCK_SIZE; i++) {
```
[the remark this answers is not in the document]
Comment (ssharks) [reply tests/net/socket/tcp/src/main.c]: Done

File: tests/net/socket/tcp/src/main.c (hunk 8 above)
Comment (ssharks) [reply tests/net/socket/tcp/src/main.c]: Indeed obsolete code

File: tests/net/socket/tcp/src/main.c (hunk 9 above)
Comment (ssharks) [reply tests/net/socket/tcp/src/main.c]: Good suggestion, done

File: drivers/net/Kconfig [hunk 12]
```diff
@@ -183,6 +183,11 @@ menuconfig NET_LOOPBACK
 
 if NET_LOOPBACK
 
+config NET_LOOPBACK_SIMULATE_PACKET_DROP
+	bool "Controlable packet drop"
+	help
+	  Enable interface to have a controlable packet drop rate
```
Comment (jukkar) [inline drivers/net/Kconfig:189]: It would be nice to emphasize that this is only used by testing and should not be enabled by end user.

File: tests/net/socket/tcp/src/main.c [hunk 13]
```diff
@@ -234,6 +235,114 @@ void test_v6_send_recv(void)
 	k_sleep(TCP_TEARDOWN_TIMEOUT);
 }
 
+/* Test the stack behavior with a resonable sized block data, be sure to have multiple packets */
+#define TEST_BLOCK_SIZE 60000
+
+static uint8_t testBlockTcp[TEST_BLOCK_SIZE];
+static uint8_t testBlockTcpRx[TEST_BLOCK_SIZE + 10];
+
+#define TCP_SERVER_STACK_SIZE 4096
```
Comment (jukkar) [inline tests/net/socket/tcp/src/main.c:244]: This and the TEST_BLOCK_SIZE will most definitely mean that the unit test cannot be run on all the devices it is currently running by twister. I just wonder could we make these numbers a bit smaller?

File: include/net/loopback.h [hunk 14]
```diff
@@ -0,0 +1,31 @@
+/** @file
+ * @brief Loopback control interface
+ */
+
+/*
+ * Copyright (c) 2017 Intel Corporation
```
Comment (rlubos) [inline include/net/loopback.h:6]: 2022. Please add yourself in the copyright.

File: tests/net/socket/tcp/src/main.c [hunk 15]
```diff
@@ -245,16 +354,16 @@ void test_v4_sendto_recvfrom(void)
 	socklen_t addrlen = sizeof(addr);
 
 	prepare_sock_tcp_v4(CONFIG_NET_CONFIG_MY_IPV4_ADDR, ANY_PORT,
-			    &c_sock, &c_saddr);
```
Comment (rlubos) [inline tests/net/socket/tcp/src/main.c:248]: Please don't introduce unnecessary changes.

File: tests/net/socket/tcp/src/main.c [hunk 16]
```diff
@@ -245,16 +354,16 @@ void test_v4_sendto_recvfrom(void)
...
+				&c_sock, &c_saddr);
 	prepare_sock_tcp_v4(CONFIG_NET_CONFIG_MY_IPV4_ADDR, SERVER_PORT,
-			    &s_sock, &s_saddr);
```
Comment (rlubos) [inline tests/net/socket/tcp/src/main.c:250]: Also here.

File: tests/net/socket/tcp/src/main.c [hunk 17]
```diff
@@ -245,16 +354,16 @@ void test_v4_sendto_recvfrom(void)
...
+				&s_sock, &s_saddr);
 
 	test_bind(s_sock, (struct sockaddr *)&s_saddr, sizeof(s_saddr));
 	test_listen(s_sock);
 
 	test_connect(c_sock, (struct sockaddr *)&s_saddr, sizeof(s_saddr));
 	test_sendto(c_sock, TEST_STR_SMALL, strlen(TEST_STR_SMALL), 0,
-		    (struct sockaddr *)&s_saddr, sizeof(s_saddr));
+			(struct sockaddr *)&s_saddr, sizeof(s_saddr));
```
Comment (rlubos) [inline tests/net/socket/tcp/src/main.c:366]: And here.

**Comment by rlubos** [comment]
Could you also split the changes in the loopback interface and the actual test into separate commits?

File: tests/net/socket/tcp/src/main.c (hunk 17 above)
Comment (ssharks) [reply tests/net/socket/tcp/src/main.c:366]: Same here, reverted

File: include/net/loopback.h (hunk 14 above)
Comment (ssharks) [reply include/net/loopback.h:6]: You're totally right, lazy copy paste action

File: tests/net/socket/tcp/src/main.c (hunk 13 above)
Comment (ssharks) [reply tests/net/socket/tcp/src/main.c:244]: I'll give it a try, the challenge that is needs to be a certain size to trigger the issue...

Comment (rlubos) [reply tests/net/socket/tcp/src/main.c:244]: Yes, but you can just keep sending the same buffer over and over, the data doesn't need to be unique (of course it would still be good to verify data integrity after each iteration).

Comment (ssharks) [reply tests/net/socket/tcp/src/main.c]: I have significantly reduced this, the only thing left is 2048 bytes of stack size, the issues still reproduces without the fixes

**Comment by ssharks** [comment]
Thanks for your review input, this helps to get the test quality up

**Review by jukkar — APPROVED** [review]

**Review by rlubos — APPROVED** [review]
